File: src/main/java/com/adven/concordion/extensions/exam/commands/ExamVerifyCommand.java
package com.adven.concordion.extensions.exam.commands;

import com.adven.concordion.extensions.exam.html.Html;
import org.concordion.api.Element;
import org.concordion.api.ResultRecorder;
import org.concordion.api.listener.AssertEqualsListener;
import org.concordion.api.listener.AssertFailureEvent;
import org.concordion.api.listener.AssertSuccessEvent;
import org.concordion.internal.util.Announcer;

import static org.concordion.api.Result.FAILURE;
import static org.concordion.api.Result.SUCCESS;

public class ExamVerifyCommand extends ExamCommand {
    private Announcer<AssertEqualsListener> listeners = Announcer.to(AssertEqualsListener.class);

    public ExamVerifyCommand(String name, String tag, AssertEqualsListener listener) {
        super(name, tag);
        listeners.addListener(listener);
    }

    protected void success(ResultRecorder resultRecorder, Element element) {
        resultRecorder.record(SUCCESS);
        listeners.announce().successReported(new AssertSuccessEvent(element));
    }

    protected void success(ResultRecorder resultRecorder, Html element) {
        success(resultRecorder, element.el());
    }

    protected void failure(ResultRecorder resultRecorder, Element element, Object actual, String expected) {
        resultRecorder.record(FAILURE);
        listeners.announce().failureReported(new AssertFailureEvent(element, expected, actual));
    }

    protected void failure(ResultRecorder resultRecorder, Html element, Object actual, String expected) {
        failure(resultRecorder, element.el(), actual, expected);
    }
}

File: src/main/java/com/adven/concordion/extensions/exam/core/commands/ExamVerifyCommand.kt
package com.adven.concordion.extensions.exam.core.commands

import com.adven.concordion.extensions.exam.core.html.Html
import org.concordion.api.Element
import org.concordion.api.Result.FAILURE
import org.concordion.api.Result.SUCCESS
import org.concordion.api.ResultRecorder
import org.concordion.api.listener.AssertEqualsListener
import org.concordion.api.listener.AssertFailureEvent
import org.concordion.api.listener.AssertSuccessEvent
import org.concordion.internal.util.Announcer

open class ExamVerifyCommand(name: String, tag: String, listener: AssertEqualsListener) : ExamCommand(name, tag) {
    private val listeners = Announcer.to(AssertEqualsListener::class.java)

    init {
        listeners.addListener(listener)
    }

    protected fun success(resultRecorder: ResultRecorder, element: Element) {
        resultRecorder.record(SUCCESS)
        listeners.announce().successReported(AssertSuccessEvent(element))
    }

    protected fun success(resultRecorder: ResultRecorder, element: Html) = success(resultRecorder, element.el())

    protected fun failure(resultRecorder: ResultRecorder, element: Element, actual: Any, expected: String) {
        resultRecorder.record(FAILURE)
        listeners.announce().failureReported(AssertFailureEvent(element, expected, actual))
    }

    protected fun failure(resultRecorder: ResultRecorder, element: Html, actual: Any, expected: String) =
            failure(resultRecorder, element.el(), actual, expected)

    protected fun ResultRecorder.pass(element: Html) = success(this, element)

    protected fun ResultRecorder.failure(element: Html, actual: String, expected: String) =
            failure(this, element, actual, expected)

    protected fun ResultRecorder.check(root: Html, actual: String, expected: String, test: (String, String) -> Boolean) =
            if (test(actual, expected)) this.pass(root) else this.failure(root, actual, expected)
}